### ISSUE TITLE
Refactor Adam and AdamW by abstracting out common code

### ISSUE THREAD

**Issue by janeyx99** [pytorch/pytorch#104899]
AdamW differs with Adam only in the weight_decay handling. Everything else is the same. We should reuse code instead of hosting the same exact logic in two places.

One way to do this is to have certain functions in adam.py that can be accessed in adamw.py. The main work that would need to be done would be consolidating the forloop and foreach implementations. The fused implementation already takes advantage of the commonality. 

cc @vincentqb @jbschlosser @albanD @crcrpar

**Comment by vadimkantorov** [comment]
overall, I think there is value in fully self-contained codes as they can be good basis for modification for study / research (like when people do single-file impls in RL)

but maybe Adam can just be united with AdamW (and just be different options) or just one inherits from another?

**Comment by tfsingh** [comment]
Hi @janeyx99 , can you assign this issue and [this one](https://github.com/pytorch/pytorch/issues/101096) to me? Have some more time now and plan on tackling both of these

**Comment by janeyx99** [comment]
@tfsingh done!

**Comment by tfsingh** [comment]
Hi @janeyx99 @vadimkantorov , finally coming back to this and plan on opening a PR this weekend. Before I do that, I wanted to check if you both were on-board with having AdamW inherit from Adam as this is my current plan.

**Comment by vadimkantorov** [comment]
(just to point out, I'm not part of dev team and not very competent about optimizers' path forward)

**Comment by tfsingh** [comment]
Ah okay, I've seen you on many threads in the past so I made a hasty assumption. Your input makes complete sense to me though, I'll do more research to improve my contextual understanding of these specific optimizers but there seems to be a lot of overlap on face.

**Comment by janeyx99** [comment]
Yes! We are still interested in doing this. We can do something similar to "NAdamW/RAdamW" where there's a kwarg on Adam `decoupled_weight_decay`, and consolidate most of the code onto Adam. This will be a little more complex than NAdam and RAdam because we have fused implementations. It is okay to branch in the fused_adam code into fused_adamw if decoupled_weight_decay=True though.

It is also tantamount to still maintain the public API of torch.optim.AdamW, as many people will use it. However, there should be much less code in adamw.py after this.

**Comment by qqaatw** [comment]
I'd assume that this issue has not been resolved, as the relevant PRs are either closed or reverted?

**Comment by janeyx99** [comment]
This issue is still up for grabs yes, though @tfsingh has  done a lot of the foundational work for it in his PRs.

**Comment by qqaatw** [comment]
I would let @tfsingh get his PR merged as he has done a lot work. Was just a kind reminder that this issue has not been resolved :)

**Comment by tfsingh** [comment]
Going to be a while until I can tackle this (2 weeks at minimum, but I
should be more free then)

**Comment by ad8e** [comment]
Reading the commit, does this mean existing users of torch.AdamW will default to `decoupled_weight_decay=True`?

**Comment by ad8e** [comment]
In the code, `decoupled_weight_decay` refers to the meaning of "apply WD to the params instead of adding to the gradient", but modern usage of "decoupled" weight decay typically means "weight decay is not scaled by LR". For example, https://fabian-sp.github.io/posts/2024/02/decoupling/ uses this meaning.

https://arxiv.org/pdf/2407.05872 follows this: `In decoupled weight decay, the weight decay is not scaled by the base learning rate`

**Comment by janeyx99** [comment]
Haha your comment took me down a rabbit hole that I'm still currently miring through. I've gotten to the point where I've rediscovered it is a known concern from this old comment: (https://github.com/pytorch/pytorch/pull/4429#issuecomment-376169085).

I'll read through that whole thread before feeling 100% confident about what we should do, though at the very least I admit we can improve our docs so that's a clear next step.

**Comment by ad8e** [comment]
The current default torch behavior is fine, where weight decay is "coupled" to LR. People aware that torch has coupled weight decay can work around it if they want. The name "decoupled" in the adamw code would cause confusion (hence my initial comment, now deleted), but that's the only issue.

In retrospect, coupled weight decay (= tied to LR) is a better default, despite the papers I linked (and other papers too). The most obvious problems with decoupled weight decay are with LR schedulers and per-tensor LRs. As an example, if LR=0 at the end of a training run, then a few thousand steps with decoupled weight decay would cause model collapse, since the gradient-based updates are all 0 but the weight decay keeps shrinking the weights.

**Comment by janeyx99** [comment]
Yea, it's been ingrained so long it would be quite hard to change the implementation, even if we wanted to 😛 

Though I'd expect the issue to come when LRScheduler is used in general, no matter if there's per-Tensor lr or not. In those cases, people would have to manually schedule the weight decay too, but that's not an obvious thing to do today. Maybe it is time to resolve https://github.com/pytorch/pytorch/issues/22343
